terraform/node_module_removed.go

```diff
@@ -0,0 +1,71 @@
+package terraform
+
+import (
+	"fmt"
+	"log"
+	"reflect"
+)
+
+// NodeModuleRemoved represents a module that is no longer in the
+// config.
+type NodeModuleRemoved struct {
+	PathValue []string
+}
+
+func (n *NodeModuleRemoved) Name() string {
+	return fmt.Sprintf("%s (removed)", modulePrefixStr(n.PathValue))
+}
+
+// GraphNodeSubPath
+func (n *NodeModuleRemoved) Path() []string {
+	return n.PathValue
+}
+
+// GraphNodeEvalable
+func (n *NodeModuleRemoved) EvalTree() EvalNode {
+	return &EvalOpFilter{
+		Ops: []walkOperation{walkRefresh, walkApply, walkDestroy},
+		Node: &EvalDeleteModule{
+			PathValue: n.PathValue,
+		},
+	}
+}
+
+// EvalDeleteModule is an EvalNode implementation that removes an empty module
+// entry from the state.
+type EvalDeleteModule struct {
+	PathValue []string
+}
+
+func (n *EvalDeleteModule) Eval(ctx EvalContext) (interface{}, error) {
+	state, lock := ctx.State()
+	if state == nil {
+		return nil, nil
+	}
+
+	// Get a write lock so we can access this instance
+	lock.Lock()
+	defer lock.Unlock()
+
+	// Make sure we have a clean state
+	// Destroyed resources aren't deleted, they're written with an ID of "".
+	state.prune()
+
+	// find the module and delete it
+	for i, m := range state.Modules {
+		if reflect.DeepEqual(m.Path, n.PathValue) {
+			if !m.Empty() {
+				// a targeted apply may leave module resources even without a config,
+				// so just log this and return.
+				log.Printf("[DEBUG] cannot remove module %s, not empty", modulePrefixStr(n.PathValue))
+				break
+			}
+			tail := len(state.Modules) - 1
+			state.Modules[i] = state.Modules[tail]
+			state.Modules = state.Modules[:tail]
```

Comment:
Am I understanding correctly that this moves whichever module is last into the position of the module being deleted?

If so, while indeed the ordering here isn't that important, Terraform Enterprise shows diffs between states and so this reordering may cause some confusion for that diffing engine.

Could we instead retain the relative ordering of all of the other modules by shifting up everything after this point?

```go
copy(state.Modules[i], state.Modules[i+1:])
state.Modules = state.modules[len(state.Modules)-1]
```

I think there may also be a bug here where we skip over one module for each one we remove, since after we've finished closing the gap here (whichever way we do it) there's a different object at `i` which we'll never see, because the next iteration will have already incremented `i`. Perhaps safer and easier to understand here if we were to just construct a new slice altogether and append to it, rather than trying to close this up in-place... this slice should never be large, so copying shouldn't hurt.

Reply:
Yes, that's the most efficient delete pattern.

Good point, even though these may be sorted again later, I'll change this to preserve order just in case. The standard delete preserving order looks like:
```
state.Modules = append(state.Modules[:i], state.Modules[i+1:]...)
```

Reply:
Updated. There no counting bug, since EvalDeleteModule only ever removes a single module.

Reply:
Ahh, I see... the `break` in there to make it it stop after one action. Thanks for the clarification!